backend/src/test/scala/com/softwaremill/adopttapir/starter/StarterServiceITTest.scala

```diff
@@ -247,14 +248,65 @@ case class Service(tempDir: better.files.File) {
   }
 }
 
+case class SbtService(tempDir: better.files.File) extends Service {
```

Comment:
As it is used only by SbtService object it can be hidden there: 

```
object SbtService {
  def apply(tempDir: better.files.File): IO[SbtService] = {
    IO.blocking(SbtService(tempDir))
  }

  private case class SbtService(tempDir: better.files.File) extends Service {
    override protected val portPattern = new Regex("^(?:\\[info\\] )(?:Go to |Server started at )http://localhost:(\\d+).*")

    override protected lazy val process: SubProcess = {
      os.proc(
        "sbt",
        "-no-colors",
        // start in forked mode so that process input can be forwarded and process waits before closing otherwise `StdIn.readLine` will exit immediately
        "set run / fork := true",
        // forward std input to forked process - https://www.scala-sbt.org/1.x/docs/Forking.html#Configuring+Input
        "set run / connectInput := true",
        ";compile ;test ;run"
      ).spawn(cwd = os.Path(tempDir.toJava), env = Map("http.port" -> "0"), mergeErrIntoOut = true)
    }
  }
```

Reply:
Same is true for Scala Cli

Reply:
Or even

object Service {
    def sbt (starterDetails: StarterDetails): IO[Service] = ??? 
    def scalaCLI(starterDetails: StarterDetails): IO[Service] = ??? 
}

so you will have one object responsible to provide different implementations

Reply:
I have introduced object `Service` that hides selection of implementation depending on a builder type
```
object Service {
  import Timeouts.waitForScalaCliCompileAndUnitTest

  def apply(builder: Builder, tempDir: better.files.File): IO[Service] = {
    builder match {
      case Builder.Sbt      => IO.blocking(SbtService(tempDir))
      case Builder.ScalaCli => IO.blocking(ScalaCliService(tempDir)).timeoutAndForget(waitForScalaCliCompileAndUnitTest)
    }
  }

  private case class SbtService(tempDir: better.files.File) extends Service {
  ...
  }

  private case class ScalaCliService(tempDir: better.files.File) extends Service {
  ...
  }
```